### PR TITLE
Improve assert message for invalid FE indices in DoFAccessor

### LINKED ISSUE #13418: Error when extract face dofs of face boardering `FE_Nothing`
I thought the following would work, but it results in an exception. Is this a bug or not?

Set up a triangulation with 2 cells:

`|cell 0|cell 1|`

 and a `DoFHandler` with elements over the triangulation:

` |FE_Q|FE_Nothing|`

 Try to extract the face dofs of cell 0 at the shared face.

```c++
#include <deal.II/dofs/dof_handler.h>

#include <deal.II/fe/fe_nothing.h>
#include <deal.II/fe/fe_q.h>

#include <deal.II/grid/grid_generator.h>

#include <vector>

using namespace dealii;

template <int dim>
void
make_two_element_mesh(Triangulation<dim> &triangulation)
{
  const Point<dim> lower_left;
  Point<dim>       upper_right;

  const unsigned int        n_cells = 2;
  std::vector<unsigned int> repetitions;
  upper_right[0] = n_cells;
  repetitions.push_back(n_cells);
  for (unsigned int d = 1; d < dim; ++d)
    {
      upper_right[d] = 1;
      repetitions.push_back(1);
    }

  GridGenerator::subdivided_hyper_rectangle(triangulation,
                                            repetitions,
                                            lower_left,
                                            upper_right);
}



int
main()
{
  const int dim = 1;

  Triangulation<dim> triangulation;
  make_two_element_mesh(triangulation);

  hp::FECollection<dim> fe_collection;
  fe_collection.push_back(FE_Nothing<dim>());
  fe_collection.push_back(FE_Q<dim>(1));

  DoFHandler<dim> dof_handler(triangulation);
  dof_handler.begin_active()->set_active_fe_index(1);
  dof_handler.distribute_dofs(fe_collection);

  const auto cell = dof_handler.begin_active();

  std::vector<types::global_dof_index> dof_indices(
    fe_collection[1].dofs_per_cell);

  // Get the dofs at the shared face.
  cell->face(1)->get_dof_indices(dof_indices);
}
```

```
--------------------------------------------------------
An error occurred in line <296> of file </home/simon/bin/dealii/include/deal.II/dofs/dof_accessor.templates.h> in function
    static void dealii::internal::DoFAccessorImplementation::Implementation::process_dof_index(const dealii::DoFHandler<dim, spacedim>&, unsigned int, unsigned int, unsigned int, unsigned int, const std::integral_constant<int, structdim>&, GlobalIndexType&, const DoFPProcessor&) [with int dim = 1; int spacedim = 1; int structdim = 0; GlobalIndexType = unsigned int; DoFPProcessor = dealii::internal::DoFAccessorImplementation::Implementation::get_dof_index<1, 1, 0>::<lambda(const auto:8&, auto:9&)>]
The violated condition was: 
    ptr != dof_handler.hp_object_fe_indices[structdim].begin() + dof_handler.hp_object_fe_ptr[structdim][obj_index + 1]
Additional information: 
    You are requesting an active FE index that is not assigned to any of
    the cells connected to this entity.

Stacktrace:
-----------
#0  ./example: void dealii::internal::DoFAccessorImplementation::Implementation::process_dof_index<1, 1, 0, unsigned int, dealii::internal::DoFAccessorImplementation::Implementation::get_dof_index<1, 1, 0>(dealii::DoFHandler<1, 1> const&, unsigned int, unsigned int, unsigned int, unsigned int, std::integral_constant<int, 0> const&)::{lambda(auto:1 const&, auto:2&)#1}>(dealii::DoFHandler<1, 1> const&, unsigned int, unsigned int, unsigned int, unsigned int, std::integral_constant<int, 0> const&, unsigned int&, dealii::internal::DoFAccessorImplementation::Implementation::get_dof_index<1, 1, 0>(dealii::DoFHandler<1, 1> const&, unsigned int, unsigned int, unsigned int, unsigned int, std::integral_constant<int, 0> const&)::{lambda(auto:1 const&, auto:2&)#1} const&)
#1  ./example: unsigned int dealii::internal::DoFAccessorImplementation::Implementation::get_dof_index<1, 1, 0>(dealii::DoFHandler<1, 1> const&, unsigned int, unsigned int, unsigned int, unsigned int, std::integral_constant<int, 0> const&)
#2  ./example: dealii::DoFAccessor<0, 1, 1, false>::get_dof_indices(std::vector<unsigned int, std::allocator<unsigned int> >&, unsigned int) const
#3  ./example: main
```
--------------------------------------------------------

### PR DESCRIPTION
**Pull request by peterrum**
closes #13418

FYI @simonsticko

### CODE REVIEW

**Review by kronbichler — APPROVED** [review]
Looks good. Can you point me to the message that now comes up?

**Comment by peterrum** [comment]
@kronbichler 

https://github.com/dealii/dealii/blob/8755d7f009e5ca026beafc31e1bdaba7b07ee8ca/include/deal.II/dofs/dof_accessor.templates.h#L235-L237

**Review by marcfehling — APPROVED** [review]

**Review by kronbichler — APPROVED** [review]

**Review by masterleinad — APPROVED** [review]

File: include/deal.II/dofs/dof_accessor.templates.h [hunk 1]
```diff
@@ -197,6 +197,49 @@ namespace internal
...
+          // Otherwise: If anything other than the default is provided by
+          // the caller, then we should take just that. As an exception, if
+          // we are on a cell (rather than a face/edge/vertex), then we know
+          // that there is only one active fe index on this cell and we can
+          // use that:
+          if ((dim == structdim) &&
+              (fe_index == DoFHandler<dim, spacedim>::invalid_fe_index))
+            {
+              return cell.nth_active_fe_index(0);
+            }
```
Comment (marcfehling) [inline include/deal.II/dofs/dof_accessor.templates.h:233]: Maybe we can be extra cautious here.
```suggestion
          if ((dim == structdim) &&
              (fe_index == DoFHandler<dim, spacedim>::invalid_fe_index))
            {
              Assert(cell.n_active_fe_indices() == 1, ExcInternalError());
              return cell.nth_active_fe_index(0);
            }
```
or:
```suggestion
          if ((cell.n_active_fe_indices() == 1) &&
              (fe_index == DoFHandler<dim, spacedim>::invalid_fe_index))
            {
              return cell.nth_active_fe_index(0);
            }
```
or my favorite:
```suggestion
          if (cell.n_active_fe_indices() == 1)
            {
              return cell.nth_active_fe_index(0);
            }
```

What do you think?

File: include/deal.II/dofs/dof_accessor.templates.h [hunk 2]
```diff
@@ -197,6 +197,49 @@ namespace internal
 {
   namespace DoFAccessorImplementation
   {
+    /**
+     * Convert an FE index that might contain the right value but also
+     * invalid_fe_index to a right value if needed/possible.
+     */
+    template <int structdim, int dim, int spacedim, bool level_dof_access>
+    unsigned int
+    get_fe_index(
```
Comment (bangerth) [inline include/deal.II/dofs/dof_accessor.templates.h:206]: This kinds of functions are often called `*_or_something()`. A couple of recent additions to the standard are https://en.cppreference.com/w/cpp/utility/optional/value_or and https://en.cppreference.com/w/cpp/utility/optional/or_else

My suggestion would be to call this function `get_fe_index_or_default()` instead.

File: include/deal.II/dofs/dof_accessor.templates.h (hunk 1 above)
Comment (bangerth) [reply include/deal.II/dofs/dof_accessor.templates.h:233]: I like option 2. Option three would need an assertion that `fe_index` either equals `invalid_fe_index` or the correct index.

**Review by bangerth — COMMENTED** [review]
Nice simplification!

File: include/deal.II/dofs/dof_accessor.templates.h (hunk 1 above)
Comment (peterrum) [reply include/deal.II/dofs/dof_accessor.templates.h]: I went with option 1. This is an optimization for cells so let's keep it this way.

File: include/deal.II/dofs/dof_accessor.templates.h (hunk 2 above)
Comment (peterrum) [reply include/deal.II/dofs/dof_accessor.templates.h]: I have rename it to `get_fe_index_or_default()`.

**Review by bangerth — APPROVED** [review]

**Pull request merged**